PG-1863 Do not try to delete keys or log WAL for temporary tables

jeltz opened the pull request:
We forgot to have a check against trying to delete leftover SMGR keys for temporary tables which is a useless operation since they are in stored in memory.

Additionally we forgot to prevent WAL from being written when creating or removing a key in `smgrcreate()` for temporary tables.

jeltz committed:
PG-1863 Do not try to delete keys or log WAL for temporary tables

We forgot to have a check against trying to delete leftover SMGR keys
for temporary tables which is a useless operation since they are store
in memory.

Additionally we forgot to prevent WAL from being written when creating
or removing a key in smgrcreate() for temporary tables.

AndersAstrand reviewed (changes requested):
Why do we prefer "remove" over "delete" here? We actually delete the keys, not just remove them from our collection of keys while they still exist elsewhere. Right?

@artemgavrilov any comments?

artemgavrilov commented:
> @artemgavrilov any comments?

Yep, looks a little bit messy. tde_smgr_log_`create`_key but XLOG_TDE_`ADD`_RELATION_KEY, tde_smgr_log_`remove`_leftover_key but XLOG_TDE_`DELETE`_RELATION_KEY.

As for my preference it should be `delete` for things that we `create`d and `remove` for things that we `add`ed

jeltz committed:
PG-1863 Consistently use create/delete for keys

We used a mix of create, add, delete and remove. We still use free and
save in pg_tde_tdemap.c but that is soemthing we can fix later.

jeltz commented:
Fixed naming to make it slightly more consistent. I did not fix the issues in `pg_tde_tdemap.c`.

AndersAstrand reviewed (changes requested):
I'm not sure why we first rename functions from "delete" to "remove" to then name them back.

And with regards to the principal keys XLOG_ constant I think "add" is appropriate since we don't create anything just add "knowledge" of a key that exists elsewhere (in the KMS).

jeltz commented:
Add does not just add knowledge, it generates a new key too.

jeltz commented:
I had planned to fix the renaming back and forth later after the bikeshedding is finished.

AndersAstrand commented:
> Add does not just add knowledge, it generates a new key too.

I'm confused. We only _create_ principal keys with the pg_tde_create_key_using_..._()` call, right? And those are then "added" using `pg_set_key_using_..._()` calls which emits the WAL record in question.

The change for the relation keys from add to create looks fine to me :).

AndersAstrand reviewed (changes requested):
I think the second commit should be done first, or as a separate PR.

I think the "_leftover_" clarification in the names from the second commit should be done in the currently first commit though, because it's unrelated to renaming the XLOG_ constants.

AndersAstrand reviewed (approved):
Looks good!

I'm sorry for the confusion. This is why I shouldn't have pushed cleaning the naming up until after the patch I did for adding the WAL record to remove leftover keys. Sorry about that.
